File: CS/ImageMemoEdit/ExtendedEditor/MyMemoEditViewInfo.cs
using System;
using System.Collections.Generic;
using System.Linq;
using DevExpress.XtraEditors.Repository;
using DevExpress.XtraEditors.ViewInfo;
using System.Drawing;
using DevExpress.Utils.Drawing;
using DevExpress.XtraEditors.Drawing;

namespace EditorDescendant
{
    public class MyMemoEditViewInfo : MemoEditViewInfo, IHeightAdaptable
    {
        int IHeightAdaptable.CalcHeight(GraphicsCache cache, int width)
        {
            width -= ((MyRepositoryItemMemoEdit)Item).ContentImageSize.Width;
            BorderObjectInfoArgs info = new BorderObjectInfoArgs(cache);
            info.Bounds = new Rectangle(0, 0, width, 100);
            Rectangle textRect = BorderPainter.GetObjectClientRectangle(info);
            if (!(BorderPainter is EmptyBorderPainter) && !(BorderPainter is InplaceBorderPainter))
                textRect.Inflate(-1, -1);
            string text = string.Empty;
            if (Item.LinesCount == 0)
            {
                text = DisplayText;
                if (text != null && text.Length > 0)
                {
                    char lastChar = text[text.Length - 1];
                    if (lastChar == 13 || lastChar == 10) text += "W";
                }
            }
            else
            {
                for (int i = 0; i < Item.LinesCount; i++)
                    text += (string.IsNullOrEmpty(text) ? "" : Environment.NewLine) + "W";
            }
            int height1 = CalcTextSizeCore(cache, text, textRect.Width).Height + 1;
            return (height1 + 100 - textRect.Bottom) + 1;
        }

        public MyMemoEditViewInfo(RepositoryItem item)
            : base(item)
        { }

        protected override Rectangle CalcMaskBoxRect(Rectangle content, ref Rectangle contextImageBounds)
        {
            Rectangle r = base.CalcMaskBoxRect(content, ref contextImageBounds);
            int width = IconRect.Width;
            r.Width = r.Width - width;
            r.X = r.X + width;
            return r;
        }
        
        protected override void Assign(BaseControlViewInfo info)
        {
            base.Assign(info);
            MyMemoEditViewInfo be = info as MyMemoEditViewInfo;
            if (be == null) return;
            this.fIconRect = be.fIconRect;
        }

        protected override Size CalcClientSize(Graphics g)
        {
            Size size = base.CalcClientSize(g);
            size.Width += this.Item.ContentImageSize.Width;
            return size;
        }
        
        protected override void CalcContentRect(Rectangle bounds)
        {
            Rectangle r = Rectangle.Empty;
            this.fIconRect = CalcIconRect(ContentRect);
            base.CalcContentRect(bounds);
            this.fMaskBoxRect = CalcMaskBoxRect(bounds, ref r);
        }

        protected virtual Rectangle CalcIconRect(Rectangle content)
        {
            Rectangle r = fMaskBoxRect;
            r.Size = Item.ContentImageSize;
            r.Location = content.Location;
            return r;
        }

        public virtual new MyRepositoryItemMemoEdit Item
        {
            get { return base.Item as MyRepositoryItemMemoEdit; }
        }

        public Image GetImage() 
        {
            return Item.GetImage();
        }

        Rectangle fIconRect;

        protected internal virtual Rectangle IconRect
        {
            get { return fIconRect; }
            set { fIconRect = value; }
        }
    }
}

File: VB/ImageMemoEdit/ExtendedEditor/MyMemoEditViewInfo.vb
Imports System
Imports DevExpress.XtraEditors.Repository
Imports DevExpress.XtraEditors.ViewInfo
Imports System.Drawing
Imports DevExpress.Utils.Drawing
Imports DevExpress.XtraEditors.Drawing

Namespace EditorDescendant

    Public Class MyMemoEditViewInfo
        Inherits MemoEditViewInfo
        Implements IHeightAdaptable

        Private Overloads Function CalcHeight(ByVal cache As GraphicsCache, ByVal width As Integer) As Integer Implements IHeightAdaptable.CalcHeight
            width -= CType(Item, MyRepositoryItemMemoEdit).ContentImageSize.Width
            Dim info As BorderObjectInfoArgs = New BorderObjectInfoArgs(cache)
            info.Bounds = New Rectangle(0, 0, width, 100)
            Dim textRect As Rectangle = BorderPainter.GetObjectClientRectangle(info)
            If Not(TypeOf BorderPainter Is EmptyBorderPainter) AndAlso Not(TypeOf BorderPainter Is InplaceBorderPainter) Then textRect.Inflate(-1, -1)
            Dim text As String = String.Empty
            If Item.LinesCount = 0 Then
                text = DisplayText
                If Not Equals(text, Nothing) AndAlso text.Length > 0 Then
                    Dim lastChar As Char = text(text.Length - 1)
                    If lastChar.Equals(13) OrElse lastChar.Equals(10) Then text += "W"
                End If
            Else
                For i As Integer = 0 To Item.LinesCount - 1
                    text += If(String.IsNullOrEmpty(text), "", Environment.NewLine) & "W"
                Next
            End If

            Dim height1 As Integer = CalcTextSizeCore(cache, text, textRect.Width).Height + 1
            Return height1 + 100 - textRect.Bottom + 1
        End Function

        Public Sub New(ByVal item As RepositoryItem)
            MyBase.New(item)
        End Sub

        Protected Overrides Function CalcMaskBoxRect(ByVal content As Rectangle, ByRef contextImageBounds As Rectangle) As Rectangle
            Dim r As Rectangle = MyBase.CalcMaskBoxRect(content, contextImageBounds)
            Dim width As Integer = IconRect.Width
            r.Width = r.Width - width
            r.X = r.X + width
            Return r
        End Function

        Protected Overrides Sub Assign(ByVal info As BaseControlViewInfo)
            MyBase.Assign(info)
            Dim be As MyMemoEditViewInfo = TryCast(info, MyMemoEditViewInfo)
            If be Is Nothing Then Return
            fIconRect = be.fIconRect
        End Sub

        Protected Overrides Function CalcClientSize(ByVal g As Graphics) As Size
            Dim size As Size = MyBase.CalcClientSize(g)
            size.Width += Item.ContentImageSize.Width
            Return size
        End Function

        Protected Overrides Sub CalcContentRect(ByVal bounds As Rectangle)
            Dim r As Rectangle = Rectangle.Empty
            fIconRect = CalcIconRect(ContentRect)
            MyBase.CalcContentRect(bounds)
            Me.fMaskBoxRect = CalcMaskBoxRect(bounds, r)
        End Sub

        Protected Overridable Function CalcIconRect(ByVal content As Rectangle) As Rectangle
            Dim r As Rectangle = fMaskBoxRect
            r.Size = Item.ContentImageSize
            r.Location = content.Location
            Return r
        End Function

        Public Overridable Overloads ReadOnly Property Item As MyRepositoryItemMemoEdit
            Get
                Return TryCast(MyBase.Item, MyRepositoryItemMemoEdit)
            End Get
        End Property

        Public Function GetImage() As Image
            Return Item.GetImage()
        End Function

        Private fIconRect As Rectangle

        Protected Friend Overridable Property IconRect As Rectangle
            Get
                Return fIconRect
            End Get

            Set(ByVal value As Rectangle)
                fIconRect = value
            End Set
        End Property
    End Class
End Namespace
